ERCS/erc-7752.md

```diff
@@ -0,0 +1,937 @@
+    address _investorOnchainID,
+    uint256[] calldata _ids
+) external returns (bool);
+```
+
+- **Requirements**:
+  - MUST only be callable by an address with the `AgentRole`.
+  - MUST verify the identity of the investor through the Identity Registry.
+  - MUST emit the `RecoverySuccess` event upon successful recovery.
+
+### Locking and Unlocking Tokens
+
+The contract MUST allow token holders to lock and unlock their tokens:
+
+```solidity
+function lockTokens(uint256 tokenId, uint256 amount) external;
+function unlockTokens(uint256 tokenId, uint256 amount) external;
+```
+
+- **Requirements**:
+  - MUST prevent locked tokens from being transferred.
+  - MUST emit `TokensLocked` and `TokensUnlocked` events accordingly.
+
+### Events
+
+The contract MUST emit events as specified in the interface to enable off-chain monitoring and compliance:
+
+- `TokensLocked`
+- `TokensUnlocked`
+- `Approval`
+- `ApprovalForAll`
+- `Transfer`
+- `TransferValue`
+- `TokenInvalidated`
+- `URI`
+- `AgentAddedForToken`
+- `AgentRemovedForToken`
+- `UpdatedTokenInformation`
+- `IdentityRegistryAdded`
+- `ComplianceAdded`
+- `Paused`
+- `Unpaused`
+- `Mint`
+- `Burn`
+- `AddressFrozen`
+- `TokenFrozen`
+- `TokenUnfrozen`
+- `RecoverySuccess`
+
+### Access Control
+
+- The contract MUST implement role-based access control to restrict certain functions to authorized addresses (e.g., `AgentRole`).
+- Ownership and agent roles SHOULD be managed securely to prevent unauthorized access.
+
+### Compliance Requirements
+
+- The contract MUST integrate with compliance modules to enforce KYC/AML regulations.
+- Transfers MUST be checked against compliance rules before execution.
+
+### Error Handling
+
+- The contract SHOULD use standard OpenZeppelin error messages and revert reasons for consistency.
+- MUST revert transactions that violate compliance rules, are paused, or involve frozen addresses/tokens.
+
+### Backwards Compatibility
+
+This standard is not backwards compatible with [ERC-20](./eip-20.md) and [ERC-3643](./eip-3643.md) due to:
+
+- Introduction of unique `tokenId` for each issuance or transfer, differentiating tokens beyond simple fungible balances.
+- Replacement of `mint` and `burn` methods with `issue` and `cancel` in earlier drafts, although `mint` and `burn` are used in this interface.
+- Emphasis on individual holdings of securities rather than fungible token balances.
+
+### Notes
+
+- Vesting and governance mechanisms are considered separate concerns and SHOULD be addressed in companion ERCs.
+- Contracts implementing this standard SHOULD ensure that they remain within the maximum contract size limits.
+- Developers SHOULD consider gas efficiency when implementing batch operations and compliance checks.
+
+```solidity
+library Types {
+  enum Status {
+    Outstanding,
+    Burn,
+    Transferred,
+    Converted,
+    Repurchased,
+    PartiallyExercised,
+    FullyExercised,
+    Forfeited,
+    Expired
+  }
+}
+
+interface IERC7752 {
```

Comment:
I'd recommend dividing this up into different interfaces that correspond to the optional components above. Like for example:

```solidity
interface IERC7752 {
    // ...
}

interface IERC7752ForcedTransfer {
    function forcedTransfer(
      address _from,
      address _to,
      uint256 _id,
      uint256 _amount
    ) external returns (uint256 newTokenId);

    function batchForcedTransfer(
        address[] calldata _fromList,
        address[] calldata _toList,
        uint256[] calldata _ids,
        uint256[] calldata _amounts
    ) external;
}
```

Otherwise none of your optional components are actually optional because of the requirement on line 41:

> Every ERC-7752 compliant contract MUST implement the `IERC7752` interface.